# Comparisons between real and simulated data

In order to assess the efficiency of the detector, a number of monoenergetic sources were simulated in order to calculate efficiencies of capture and scattering efficiencies as a function of energy. This analysis was also performed using $^{252}$Cf, AmBe and AmLi. Real data from NPL measurements is also used to calculate an real efficiency for comparison at 144 keV and 1.2 MeV. 

In [3]:
import SimEfficiency as SE
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import RealVsSim as RVS
%matplotlib tk

from mpl_toolkits.mplot3d import Axes3D
import matplotlib.colors as colors
import matplotlib.cm as cm

In [7]:
#monos_path = '/home/nr1315/Documents/Project/Simulations/monos_efficiencies'
monos_path = '/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data'
os.chdir(monos_path)

effs, errs, stds, cap_lens, df_lens = SE.GenerateEfficiencyDataFrame(monos_path)

efficiencies_noscatter = pd.DataFrame.from_dict(effs,orient='index')
errors_noscatter = pd.DataFrame.from_dict(errs,orient='index')
spreads_noscatter = pd.DataFrame.from_dict(stds,orient='index')

efficiencies_noscatter.to_hdf(monos_path+'/eff_dataframes/efficiencies.h5',key='event')
errors_noscatter.to_hdf(monos_path+'/eff_dataframes/errors.h5',key='event')
spreads_noscatter.to_hdf(monos_path+'/eff_dataframes/spreads.h5',key='event')

effs,errs,stds,cap_lens,df_lens = None,None,None,None,None

KeyboardInterrupt: 

In [83]:
monos2_path = '/home/nr1315/Documents/Project/Simulations/monos_efficiencies'
os.chdir(monos2_path)

effs2,errs2,stds2,cap_lens2,df_lens2 = SE.GenerateEfficiencyDataFrame(monos2_path)



efficiencies = pd.DataFrame.from_dict(effs2,orient='index')
errors = pd.DataFrame.from_dict(errs2,orient='index')
spreads = pd.DataFrame.from_dict(stds2,orient='index')

efficiencies.to_hdf(monos2_path+'/eff_dataframes/efficiencies.h5',key='event')
errors.to_hdf(monos2_path+'/eff_dataframes/errors.h5',key='event')
spreads.to_hdf(monos2_path+'/eff_dataframes/spreads.h5',key='event')

effs2,errs2,stds2,cap_lens2,df_lens2 = None,None,None,None,None

In [84]:
efficiencies = pd.read_hdf('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/eff_dataframes/efficiencies.h5')
errors = pd.read_hdf('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/eff_dataframes/errors.h5')
spreads = pd.read_hdf('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/eff_dataframes/spreads.h5')

In [5]:
efficiencies_noscatter = pd.read_hdf('/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/eff_dataframes/efficiencies.h5')
errors_noscatter = pd.read_hdf('/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/eff_dataframes/errors.h5')
spreads_noscatter = pd.read_hdf('/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/eff_dataframes/spreads.h5')

In [87]:
def PlotEfficienciesComparison(eff,err,spr,ax,label,cs,fill_region=True):
    e = np.sort([i for i in eff.index if bool(~isinstance(i,str)+2)])
    ax.plot(e,100*eff.loc[e,'cap_eff'],c=cs[0],label = 'Simulated LiF captures, {}'.format(label))
    ax.errorbar(e,100*eff.loc[e,'cap_eff'],yerr = 100*err.loc[e,'cap_eff'],fmt='none',ecolor=cs[0])
    
    if fill_region:
        ax.fill_between(e,eff.loc[e,'cap_eff']-3*err.loc[e,'cap_eff'],eff.loc[e,'cap_eff']+3*err.loc[e,'cap_eff'],color=cs[1],alpha=0.3)
        ax.fill_between(e,eff.loc[e,'cap_eff']-2*err.loc[e,'cap_eff'],eff.loc[e,'cap_eff']+2*err.loc[e,'cap_eff'],color=cs[2],alpha=0.3)
        ax.fill_between(e,eff.loc[e,'cap_eff']-err.loc[e,'cap_eff'],eff.loc[e,'cap_eff']+err.loc[e,'cap_eff'],color=cs[2],alpha=0.3)

    return

In [85]:
ii = errors.query('cap_eff<0.01').index

In [27]:
os.chdir('/home/nr1315/Documents/Project/pyENDF6/pyENDF6/')
import ENDF6
os.chdir('/home/nr1315/Documents/Project/nFacetAnalysis/docs')

In [45]:
f = open('/home/nr1315/Downloads/E4R535_e4.txt')
lines = f.readlines()
sec = ENDF6.find_section(lines,MF=1,MT=3)

ValueError: ' 1  3' is not in list

In [55]:
plt.plot(x,y)
plt.xscale('log'),plt.yscale('log')

(None, None)

In [57]:
plt.plot(x,y)
plt.xscale('log')

In [32]:
x,y = ENDF6.read_table(sec)

In [39]:
plt.plot(x,y)
plt.xscale('log')

1491.994

In [ ]:
from matplotlib.ticker import MultipleLocator,LogLocator

fig = plt.figure()
ax = fig.add_subplot(111)
dose_eq_curves = pd.read_csv("/home/nr1315/Downloads/ICRP74_neutron_dose equivalent curves.csv")

PlotEfficienciesComparison(efficiencies.loc[ii],errors.loc[ii],spreads,ax,'with rescattering',cs=['xkcd:tangerine','xkcd:goldenrod','xkcd:light orange'],fill_region=False)
#PlotEfficienciesComparison(efficiencies,errors,spreads,ax,'with scattering',cs=['xkcd:tangerine','xkcd:goldenrod','xkcd:light orange'],fill_region=False)
PlotEfficienciesComparison(efficiencies_noscatter,errors_noscatter,spreads_noscatter,ax,'no rescattering',cs=['xkcd:deep blue','xkcd:light blue','xkcd:light blue'],fill_region=False)

#ax4 = ax.twinx()
#ax4.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
#ax4.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
#ax4.plot(x/1e6,y,c='g')
#ax4.set_ylabel(r'$^{6}$Li neutron capture cross section / barns',fontsize=24,c='g')
#ax4.tick_params(axis='y',labelcolor='g')
#ax4.set_ylim(0,5)
#ax4.set_yscale('log')

ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax.tick_params(axis='x',labelsize=18)

ax.tick_params(axis='y',labelsize=18)
ax.set_xlabel('Energy / MeV',fontsize=24)
ax.set_xlim(1e-8,1e2)
ax.set_xscale('log')
ax.set_ylim(0,100)
ax.legend(loc='upper left')
ax.yaxis.set_minor_locator(MultipleLocator(5))
fig.suptitle('Detection efficiency',fontsize=32)


In [7]:
import SimEventViewer as SMV

In [3]:
plotter = SMV.NewSimPlotter()
plotter.add_data('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-100.0_keV.h5','100 keV scatter')
plotter.add_data('/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/comb-100.0_keV.h5','100 keV no scatter')

Finished loading /home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-100.0_keV.h5
------------------------------
Finished loading /home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/comb-100.0_keV.h5
------------------------------


In [18]:
plotter.add_data('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-500.0_keV.h5','500 keV scatter')
plotter.add_data('/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/comb-500.0_keV.h5','500 keV no scatter')

Finished loading /home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-500.0_keV.h5
------------------------------
Finished loading /home/nr1315/Documents/Project/Simulations/monos_no_rescattering/comb-500.0_keV.h5
------------------------------


In [20]:
plotter.add_data('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-1.0_MeV.h5','1 MeV scatter')
plotter.add_data('/home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/comb-1.0_MeV.h5','1 MeV no scatter')

Finished loading /home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-1.0_MeV.h5
------------------------------
Finished loading /home/nr1315/Documents/Project/Simulations/monos_no_rescattering/new_data/comb-1.0_MeV.h5
------------------------------


In [25]:
' '.join('100 keV scatter'.split(' ')[:-1])

'100 keV'

In [34]:
def compare_scatter_noscatter(d1,d2):
    fig = plt.figure()
    ax = fig.add_subplot(121)
    ax2 = fig.add_subplot(122)

    scatter_100k_counts,scatter_100k_err = plotter.ViewCubes(d1,ax=ax,c='C0')
    no_scatter_100k_counts,no_scatter_100k_err = plotter.ViewCubes(d2,ax=ax,c='C1')
    ax.set_title('Comparison of direct and \n rescattered data',fontsize=32)
    
    diff = scatter_100k_counts - no_scatter_100k_counts
    diff_err = np.sqrt(scatter_100k_err**2 + no_scatter_100k_err**2)

    diff_r = diff/no_scatter_100k_counts
    diff_r_err = diff_r*np.sqrt((scatter_100k_err/scatter_100k_counts)**2 + (no_scatter_100k_err/no_scatter_100k_counts)**2)

    ax2.bar(np.arange(64),diff_r,yerr=diff_r_err,width=1,edgecolor='black',capsize=3,ecolor='C0',alpha=0.3)
    ax2.set_xlabel('Cube ID',fontsize=28)
    ax2.set_ylabel(r'(R$_S$ - R$_D$)/R$_D$',fontsize=28)
    ax2.tick_params(axis='both',labelsize=24)
    ax2.set_title('Fractional difference between \n direct and rescattered data',fontsize=32)
    
    ax.legend(loc='upper left',fontsize=20)
    
    fig.suptitle(' '.join(d1.split(' ')[:-1]),fontsize=26)

In [35]:
compare_scatter_noscatter('100 keV scatter','100 keV no scatter')

findfont: Font family ['serif'] not found. Falling back to DejaVu Sans.


In [36]:
compare_scatter_noscatter('500 keV scatter','500 keV no scatter')

In [37]:
compare_scatter_noscatter('1 MeV scatter','1 MeV no scatter')

In [13]:
fig = plt.figure()
ax = fig.add_subplot(121,projection='3d')
ax2 = fig.add_subplot(122,projection='3d')


plotter.View_3D('100 keV scatter',ax=ax)

plotter.View_3D('100 keV no scatter',ax=ax2)

In [15]:
fig = plt.figure()
ax = fig.add_subplot(121,projection='3d')
ax2 = fig.add_subplot(122,projection='3d')


plotter.View_3D('500 keV scatter',ax=ax)

plotter.View_3D('500 keV no scatter',ax=ax2)

In [11]:
fig = plt.figure()
ax = fig.add_subplot(111)
dose_eq_curves = pd.read_csv("/home/nr1315/Downloads/ICRP74_neutron_dose equivalent curves.csv")

PlotEfficienciesComparison(efficiencies,errors,spreads,ax,'with scattering',cs=['xkcd:tangerine','xkcd:goldenrod','xkcd:light orange'],fill_region=True)
PlotEfficienciesComparison(efficiencies_noscatter,errors_noscatter,spreads_noscatter,ax,'no rescattering',cs=['xkcd:deep blue','xkcd:light blue','xkcd:light blue'],fill_region=True)

ax4 = ax.twinx()
ax4.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
ax4.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
ax4.tick_params(axis='y',labelcolor='g')

ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax.tick_params(axis='x',labelsize=18)

ax.tick_params(axis='y',labelsize=18)
ax.set_xlabel('Energy / MeV',fontsize=24)
ax.set_xlim(1e-8,1e2)
ax.set_xscale('log')
ax.legend(loc='upper left')
fig.suptitle('Detection efficiency',fontsize=32)


Text(0.5, 0.98, 'Detection efficiency')

In [ ]:
dose_eq_curves = pd.read_csv("/home/nr1315/Downloads/ICRP74_neutron_dose equivalent curves.csv")

fig=plt.figure()
ax = fig.add_subplot(111)

ax.plot(energies, cap_eff[:-3], c = 'xkcd:tangerine', label = 'Simulated capture events', linestyle = '--')
ax.errorbar(energies, cap_eff[:-3], yerr = cap_err[:-3], fmt = 'none', ecolor = 'xkcd:tangerine')

ax.fill_between(energies,np.array(cap_eff[:-3])-3*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+3*np.array(cap_err[:-3]),color='xkcd:goldenrod',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff[:-3])-2*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+2*np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff[:-3])-np.array(cap_err[:-3]),np.array(cap_eff[:-3])+np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)

ax.plot(energies, scatt_eff[:-3], c = 'xkcd:deep blue', label = 'Simulated scatter events', linestyle = '--')
ax.errorbar(energies, scatt_eff[:-3], yerr = scatt_err[:-3], fmt = 'none', ecolor = 'xkcd:deep blue')

ax.fill_between(energies,np.array(scatt_eff[:-3])-3*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+3*np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff[:-3])-2*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+2*np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff[:-3])-np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)


#ax.plot(energies, scatt_1_eff, c = 'b', label = '1 detectable scatter')
#ax.errorbar(energies, scatt_1_eff, yerr = scatt_1_err, fmt = 'none', ecolor = 'b')

#ax.plot(energies, scatt_2_eff, c = 'xkcd:light blue', label = '2 detectable scatters')
#ax.errorbar(energies, scatt_2_eff, yerr = scatt_2_err, fmt = 'none', ecolor = 'xkcd:deep blue')

#ax.plot(energies, cap_scatt_1_eff, c = 'r', label = '1 scatter + capture')
#ax.errorbar(energies, cap_scatt_1_eff, yerr = cap_scatt_1_err, fmt = 'none', ecolor = 'r')

#ax.plot(energies, cap_scatt_2_eff, c = 'xkcd:dark red', label = '2 scatters + capture')
#ax.errorbar(energies, cap_scatt_2_eff, yerr = cap_scatt_2_err, fmt = 'none', ecolor = 'xkcd:dark red')

ax4 = ax.twinx()
ax4.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
ax4.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
ax4.tick_params(axis='y',labelcolor='g')

ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax.tick_params(axis='x',labelsize=18)

ax.tick_params(axis='y',labelsize=18)
ax.set_xlabel('Energy / MeV',fontsize=24)
ax.set_xlim(1e-8,1e2)
ax.set_xscale('log')

ax.scatter(real_es,100*np.array(real_effs),c='purple',label=r'Real capture events')
ax.errorbar(real_es,100*np.array(real_effs),yerr=100*np.array(real_errs),c='purple',fmt='none', capsize = 4)


ax.legend(loc='upper left')
fig.suptitle('Detection efficiency',fontsize=32)



In [61]:
monos_path = '/home/nr1315/Documents/Project/Simulations/monos_efficiencies/no_rescattering/NoWalls'
os.chdir(monos_path)

ef,er,st,cl,dl = SE.GenerateEfficiencyDataFrame(monos_path)

In [62]:
ef

{0.10000000000007855: {'cap_eff': 0.40455733420450835,
  'scatt_eff': 3.266906239790918e-05,
  'scatt_1_eff': 1.633453119895459e-05,
  'scatt_2_eff': 1.633453119895459e-05,
  'cap_scatt_eff': 3.266906239790918e-05,
  'cap_scatt_1_eff': 1.633453119895459e-05,
  'cap_scatt_2_eff': 1.633453119895459e-05}}

In [21]:
monos_path = '/home/nr1315/Documents/Project/Simulations/monos_efficiencies/no_rescattering'



In [2]:
def load_dataset(path):
    df = pd.read_hdf(path)
    df.reset_index(inplace=True,drop=True)
    exp=pd.concat([df['pdg'].explode(),df.loc[:,['cubeid','CapCube']].explode('cubeid')],axis=1)
    eve = exp.loc[exp.groupby(exp.index)['pdg'].count()>1]
    eve['pdg_diff'] = eve.loc[:,'pdg'].diff()
    diff = (1000010030 - 1000020040)
    finals = eve.query('pdg_diff==@diff')

    f_vals = finals.groupby(finals.index)['cubeid'].last()
    diff_ind = finals.index.unique()

    exp.loc[:,'CapCube']=0
    exp.loc[diff_ind,'CapCube']=f_vals

    df.loc[:,'CapCube']=exp.groupby(exp.index)['CapCube'].first()

    cubes = df.query('CapCube>0')['CapCube'].values
    cubes = pd.DataFrame(np.stack(np.char.split(cubes.astype(str),'0'),axis=0).astype(int)[:,[1,2,0]],columns=['X','Y','Z']) - 1
    return cubes.groupby(['Z','Y','X'])['X'].count()

    #t=pd.concat((df['edep_pvt'].explode(),df['pdg'].explode()),axis=1)
    #t['detectable_scatt']=(t['edep_pvt']>480e-3)&((t['pdg']==2112)|(t['pdg']==2212))

    #df['NScatt']=t.groupby(t.index)['detectable_scatt'].sum()

In [56]:
c_100k_noscatter = load_dataset('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/no_rescattering/NoWalls/comb-100.0_keV.h5')
c_100k_scatter = load_dataset('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-100.0_keV.h5')

In [29]:
c_500k_noscatter = load_dataset('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/no_rescattering/comb-500.0_keV.h5')
c_500k_scatter = load_dataset('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-500.0_keV.h5')

c_1M_noscatter = load_dataset('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/no_rescattering/comb-1.0_MeV.h5')
C_1M_scatter = load_dataset('/home/nr1315/Documents/Project/Simulations/monos_efficiencies/comb-1.0_MeV.h5')

In [26]:
def compare_counts_dist(fs,name):
    arr_1,arr_2 = np.zeros([4,4,4]),np.zeros([4,4,4])
    for i in fs[0].index:
        arr_1[i[::-1]] = fs[0][i]
        arr_2[i[::-1]] = fs[1][i]
    
    plt.bar(np.arange(64),(arr_1/arr_1.sum()).flatten(),width=1,alpha=0.3,label='No rescattering',edgecolor='black')
    plt.errorbar(np.arange(64),(arr_1/arr_1.sum()).flatten(),yerr = (np.sqrt(arr_1)/arr_1.sum()).flatten(),capsize=3,fmt='none')
    
    plt.bar(np.arange(64),(arr_2/arr_2.sum()).flatten(),width=1,alpha=0.3,label='With rescattering',edgecolor='black')
    plt.errorbar(np.arange(64),(arr_2/arr_2.sum()).flatten(),yerr = (np.sqrt(arr_2)/arr_2.sum()).flatten(),capsize=3,fmt='none')
    
    plt.legend(loc='upper left')
    plt.title(name,fontsize=32)
    plt.xlabel('Cube ID',fontsize=24)
    plt.ylabel('Fractional count',fontsize=24)
    plt.tick_params(axis='both',labelsize=22)

In [54]:
def compare_heatmaps(fs,name):
    arr_1,arr_2 = np.zeros([4,4,4]),np.zeros([4,4,4])
    for i in fs[0].index:
        arr_1[i[::-1]] = fs[0][i]
        arr_2[i[::-1]] = fs[1][i]
    
    fig,ax=plt.subplots(2,3,figsize=(30,20))
    
    vmin = 0
    vmax = max(np.max((arr_1/arr_1.sum()).sum(axis=0)),np.max((arr_2/arr_2.sum()).sum(axis=0)))
    ax[0,0].imshow((arr_1/arr_1.sum()).sum(axis=0).T,interpolation='nearest',cmap='viridis',origin='lower',vmin=vmin,vmax=vmax)
    ax[0,1].imshow((arr_1/arr_1.sum()).sum(axis=1),interpolation='nearest',cmap='viridis',origin='lower',vmin=vmin,vmax=vmax)
    ax[0,2].imshow((arr_1/arr_1.sum()).sum(axis=2).T,interpolation='nearest',cmap='viridis',origin='lower',vmin=vmin,vmax=vmax)
    
    ax[1,0].imshow((arr_2/arr_2.sum()).sum(axis=0).T,interpolation='nearest',cmap='viridis',origin='lower',vmin=vmin,vmax=vmax)
    ax[1,1].imshow((arr_2/arr_2.sum()).sum(axis=1),interpolation='nearest',cmap='viridis',origin='lower',vmin=vmin,vmax=vmax)
    ax[1,2].imshow((arr_2/arr_2.sum()).sum(axis=2).T,interpolation='nearest',cmap='viridis',origin='lower',vmin=vmin,vmax=vmax)
    
    ax[0,0].set_xlabel('Y',fontsize=22)
    ax[0,0].set_ylabel('Z',fontsize=22)
    
    ax[1,0].set_xlabel('Y',fontsize=22)
    ax[1,0].set_ylabel('Z',fontsize=22)
    
    ax[0,1].set_xlabel('Z',fontsize=22)
    ax[0,1].set_ylabel('Y',fontsize=22)
    
    ax[1,1].set_xlabel('Z',fontsize=22)
    ax[1,1].set_ylabel('Y',fontsize=22)
    
    ax[0,2].set_xlabel('X',fontsize=22)
    ax[0,2].set_ylabel('Y',fontsize=22)

    ax[1,2].set_xlabel('X',fontsize=22)
    ax[1,2].set_ylabel('Y',fontsize=22)
    
    ax[0,0].text(0.1,0.1,'No scatter')
    
    fig.suptitle(name,fontsize=32)

In [57]:
compare_counts_dist((c_100k_noscatter,c_100k_scatter),'100.0 keV')

In [33]:
compare_counts_dist((c_500k_noscatter,c_500k_scatter),'500.0 keV')

In [35]:
compare_counts_dist((c_1M_noscatter,C_1M_scatter),'1.0 MeV')

In [55]:
compare_heatmaps((c_100k_noscatter,c_100k_scatter),'100.0 keV')

In [17]:
arr_100k_noscatter,arr_100k_scatter = np.zeros([4,4,4]),np.zeros([4,4,4])
for i in c_100k_noscatter.index:
    arr_100k_noscatter[i[::-1]] = c_100k_noscatter[i]
    arr_100k_scatter[i[::-1]] = c_100k_scatter[i]

In [24]:
plt.bar(np.arange(64),(arr_100k_noscatter/arr_100k_noscatter.sum()).flatten(),width=1,alpha=0.3,label='No rescattering',edgecolor='black')
plt.bar(np.arange(64),(arr_100k_scatter/arr_100k_scatter.sum()).flatten(),width=1,alpha=0.3,label='With rescattering',edgecolor='black')
plt.legend(loc='upper left')

plt.errorbar(np.arange(64),(arr_100k_noscatter/arr_100k_noscatter.sum()).flatten(),yerr = (np.sqrt(arr_100k_noscatter)/arr_100k_noscatter.sum()).flatten(),capsize=3,fmt='none')
plt.errorbar(np.arange(64),(arr_100k_scatter/arr_100k_scatter.sum()).flatten(),yerr = (np.sqrt(arr_100k_scatter)/arr_100k_scatter.sum()).flatten(),capsize=3,fmt='none')

<ErrorbarContainer object of 3 artists>

In [18]:
arr_100k_noscatter,arr_100k_scatter

(array([[[  31.,   30.,   34.,   96.],
         [  66.,   86.,  102.,  126.],
         [ 268.,  356.,  402.,  401.],
         [ 552.,  895.,  936.,  886.]],
 
        [[  37.,   29.,   35.,  104.],
         [  68.,  105.,  105.,  141.],
         [ 315.,  487.,  478.,  566.],
         [ 688., 1076., 1107.,  958.]],
 
        [[  33.,   26.,   35.,  112.],
         [  64.,   96.,  106.,  158.],
         [ 310.,  526.,  554.,  487.],
         [ 751., 1078., 1136., 1061.]],
 
        [[  18.,   39.,   38.,  101.],
         [  74.,   99.,   73.,  163.],
         [ 296.,  454.,  520.,  493.],
         [ 692., 1113., 1134., 1021.]]]),
 array([[[  36.,   58.,   62.,  135.],
         [  67.,   98.,  134.,  211.],
         [ 345.,  553.,  590.,  600.],
         [ 855., 1419., 1454., 1269.]],
 
        [[  26.,   40.,   51.,  145.],
         [  99.,  124.,  168.,  285.],
         [ 492.,  767.,  844.,  847.],
         [ 996., 1683., 1739., 1629.]],
 
        [[  25.,   31.,   52.,  148.],
       

In [17]:
df_cap = df.query('CapCube>0')

In [19]:
thermal = df_cap.loc[:,['ThermalPosX','ThermalPosY','ThermalPosZ']]-np.array([0.,-1627.,1500.])

In [29]:
thermal.query('ThermalPosX>-100&ThermalPosX<100&ThermalPosY>-100&ThermalPosY<100&ThermalPosZ>-100&ThermalPosZ<100')

ThermalPosX  ThermalPosY  ThermalPosZ
0       -19.742385    28.155308    -6.846860
3       -72.128830    61.961457     0.391569
4         7.135301    96.632310    66.128988
7        60.597162    92.640048    21.213005
13       -6.675548    79.220753    62.453122
...            ...          ...          ...
93975    80.007920    85.727320   -86.712012
93980   -14.003033    74.974506    45.956944
93981   -53.937783    80.673466    60.870753
93982    -9.931091    78.254615    10.307995
93983   -13.141835    96.389082    84.184509

[28838 rows x 3 columns]

In [138]:
df.loc[thermal.query('ThermalPosX<-100|ThermalPosX>100|ThermalPosY<-100|ThermalPosY>100|ThermalPosZ<-100|ThermalPosZ>100').index]

pdgid                 energy   posx   posy      posz  \
1      [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
16     [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
49     [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
81     [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
95     [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
...       ...                    ...    ...    ...       ...   
93903  [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
93909  [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
93928  [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
93929  [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   
93956  [2112]  [0.10000000000002274]  [0.0]  [0.0]  [1500.0]   

                          momx                   momy  \
1       [-0.04288745355016443]  [-0.9967449436431216]   
16      [-0.07474446990585179]  [-0.9926153888086154]   
49       [0.03404753831209852]  [-0.9928045818417551]   
81      [0.007414839458261689]   [-0.994028430329081]   
95     [-0.048254389305477756]  [-0.9944176969106796]   
...                        ...                    ...   
93903   [0.011740883267185278]  [-0.9996740813582908]   
93909   [-0.13132344572819646]  [-0.9868935646445943]   
93928   [0.059832095256245166]  [-0.9953097603522102]   
93929    [-0.0780537259344562]  [-0.9927528199338539]   
93956   [-0.07619816148576984]  [-0.9966405056828642]   

                          momz  DepositEnergy  LiCaptureEnergy  ...  \
1       [-0.06826553779069561]       4.881205              0.0  ...   
16        [0.0955403271860347]       4.881209              0.0  ...   
49       [0.11480342943006529]       4.881205              0.0  ...   
81         [0.108869186886427]       4.881203              0.0  ...   
95       [0.09383473761680793]       4.881211              0.0  ...   
...                        ...            ...              ...  ...   
93903  [-0.022668981462866115]       4.881207              0.0  ...   
93909    [0.09378296575156553]       4.881201              0.0  ...   
93928    [0.07601711205296756]       4.881208              0.0  ...   
93929     [0.0913753488700858]       4.881206              0.0  ...   
93956  [-0.030025699298989798]       4.881208              0.0  ...   

                       cubex                  cubey                  cubez  \
1                     [1, 1]                 [4, 4]                 [2, 2]   
16              [2, 2, 2, 2]           [4, 4, 4, 4]           [4, 4, 4, 4]   
49                 [2, 3, 3]              [3, 2, 2]              [4, 4, 4]   
81           [1, 2, 2, 1, 1]        [4, 4, 3, 4, 4]        [3, 3, 4, 3, 3]   
95                 [1, 1, 1]              [4, 4, 4]              [4, 3, 3]   
...                      ...                    ...                    ...   
93903        [3, 3, 3, 3, 3]        [4, 4, 4, 4, 4]        [1, 2, 2, 2, 2]   
93909  [2, 3, 4, 2, 3, 3, 3]  [1, 1, 1, 1, 1, 1, 1]  [1, 1, 1, 2, 1, 1, 2]   
93928                 [4, 4]                 [4, 4]                 [4, 4]   
93929              [1, 1, 1]              [2, 1, 1]              [3, 3, 3]   
93956                 [1, 1]                 [4, 4]                 [2, 2]   

                                                     pdg  \
1                               [1000020040, 1000010030]   
16                  [2112, 2112, 1000020040, 1000010030]   
49                        [2112, 1000020040, 1000010030]   
81            [2112, 2112, 2112, 1000020040, 1000010030]   
95                        [2112, 1000020040, 1000010030]   
...                                                  ...   
93903         [2112, 2112, 2112, 1000020040, 1000010030]   
93909  [2112, 2112, 2112, 2112, 1000020040, 100001003...   
93928                           [1000020040, 1000010030]   
93929                     [2112, 1000020040, 1000010030]   
93956                           [1000020040, 1000010030]   

                                           

In [42]:
df.loc[49]

pdgid                                                          [2112]
energy                                          [0.10000000000002274]
posx                                                            [0.0]
posy                                                            [0.0]
posz                                                         [1500.0]
momx                                            [0.03404753831209852]
momy                                            [-0.9928045818417551]
momz                                            [0.11480342943006529]
DepositEnergy                                                4.881205
LiCaptureEnergy                                                   0.0
PvtCaptureEnergy                                                  0.0
CapPosX                                                     15.557964
CapPosY                                                  -1627.333076
CapPosZ                                                   1578.868883
ThermalPosX         

In [44]:
import ndap

In [49]:
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.cm as cm
import matplotlib.colors as colors

In [123]:
loc=np.array([0.,-1627.,1500.])
cs = 50.888

In [126]:
track = np.vstack([*df.loc[49,['XPosition','YPosition','ZPosition']].values])
track = np.roll(np.append(track,np.array([[0.], [0.], [1500.]]),axis=1),1,axis=1)

In [132]:
(track - (loc - cs)[:,np.newaxis])/cs

array([[ 1.        ,  2.01087282,  2.01307759,  2.03753585,  2.25699409,
         2.32453487,  2.31958674,  2.37312661,  3.35820822,  3.42679675,
         3.45384779,  3.44247608,  3.35820822,  2.53929392,  2.45657075,
         2.42273852,  2.42136221,  2.42015753,  2.41846061,  2.40884989,
         2.41024323,  2.41051233,  2.41070338,  2.41089442,  2.41116352,
         2.41418931,  2.42124693,  2.4223474 ,  2.42312866,  2.42402122,
         2.43465282,  2.3791123 ,  1.61487814,  1.57305923,  1.60185411,
         1.54582103,  0.98701404,  0.92340069,  0.92146957,  0.92134668,
         0.92142515,  0.92265825,  0.92286664,  0.96388504,  1.00018101,
         1.00184064,  1.00365509,  1.00533954,  1.00880365,  1.01182279,
         1.01411119,  1.00880365,  1.00365509,  1.        ,  0.99634491,
         0.99119635,  0.96107567,  0.90153529,  0.77061401,  0.97421428,
         0.98630307,  0.99119635,  0.99634491,  1.        ,  1.00365509,
         1.00880365,  1.28375303,  1.29112471,  1.2

In [117]:
track = np.vstack([*df.loc[49,['XPosition','YPosition','ZPosition']].values])
track = np.roll(np.append(track,np.array([[0.], [0.], [1500.]]),axis=1),1,axis=1)
track = (track - np.array([0.-2*50.888, -1627.-2*50.888,1500.-2*50.888 ])[:,np.newaxis])/50.888

therm_pos = df.loc[49,['ThermalPosX','ThermalPosY','ThermalPosZ']].values.astype(float)
therm_pos = (therm_pos - np.array([0.-2*50.888, -1627.-2*50.888,1500.-2*50.888 ]))/50.888

In [122]:
arr = np.zeros([4,4,4])
sc = (50.888-50)/50

p = ndap.NDArrayPlotter(arr,spacing=("even",sc,sc,sc))
p.set_alpha(0.05)

fig = plt.figure()
ax = fig.add_subplot(111,projection='3d')
fig.set_tight_layout(True)
cmap = cm.viridis
norm = colors.Normalize(vmin=0,vmax=1)
p.colors=cmap(norm(arr))
p.alphas=np.ones([4,4,4])*0.05

ax.plot(track[0],track[1],track[2])
p.render(azim=-56,elev=25,ax=ax,text=None,labels=True,space=0.5)
ax.scatter(therm_pos[0],therm_pos[1],therm_pos[2])


ax.set_xlim(-10,10)
ax.set_ylim(-10,10)
ax.set_zlim(-10,10)

(-10, 10)

In [150]:
class Wrapper:
    def __init__(self,df):
        self.index = 0
        self.cs = 50.888
        self.loc = np.array([0.,-1627.,1500.])
        self.df = df
        
    def ViewArr3D(self):
        fig = plt.figure()
        ax = fig.add_subplot(111,projection='3d')
        fig.set_tight_layout(True)
        
        norm = colors.Normalize(vmin=-0.1,vmax=0.5)
        cmap = cm.viridis
        
        arr = np.zeros([4,4,4])
        spacing = (self.cs - 50)/50
        plotter = ndap.NDArrayPlotter(arr,spacing=("even",sc,sc,sc))
        plotter.alphas = 0.05*np.ones([4,4,4])
        plotter.colors = cmap(norm(arr))
        plotter.render(azim = -56,elev = 25,ax=ax,text=None,labels=False,space=0.5)
        
        
        
        def key_event(e):
            if e.key=='right':
                self.index+=1
                self.index%=len(self.df)
                index = self.df.index[self.index]
                ax.cla()
                
                plotter.render(azim = -56,elev = 25,ax=ax,text=None,labels=False,space=0.5)
                
                track = np.vstack([*self.df.loc[index,['XPosition','YPosition','ZPosition']].values])
                track = np.roll(np.append(track,np.array([[0.],[0.],[1500.]]),axis=1),1,axis=1)
                track = (track - (self.loc - 2*self.cs)[:,np.newaxis])/self.cs
                
                ax.plot(track[0],track[1],track[2])
                
                therm_pos = df.loc[index,['ThermalPosX','ThermalPosY','ThermalPosZ']].values.astype(float)
                therm_pos = (therm_pos - (self.loc - 2*self.cs))/self.cs
                
                ax.scatter(therm_pos[0],therm_pos[1],therm_pos[2],c='r')
                
                ax.set_xlim(-10,10)
                ax.set_ylim(-10,10)
                ax.set_zlim(-10,10)
                
                fig.canvas.draw()
                
            if e.key=='left':
                self.index-=1
                self.index%=len(self.df)
                index = self.df.index[self.index]
                ax.cla()
                
                plotter.render(azim = -56,elev = 25,ax=ax,text=None,labels=False,space=0.5)
                
                track = np.vstack([*self.df.loc[index,['XPosition','YPosition','ZPosition']].values])
                track = np.roll(np.append(track,np.array([[0.],[0.],[1500.]]),axis=1),1,axis=1)
                track = (track - (self.loc - 2*self.cs)[:,np.newaxis])/self.cs
                
                ax.plot(track[0],track[1],track[2])
                
                therm_pos = df.loc[index,['ThermalPosX','ThermalPosY','ThermalPosZ']].values.astype(float)
                therm_pos = (therm_pos - (self.loc - 2*self.cs))/self.cs
                
                ax.scatter(therm_pos[0],therm_pos[1],therm_pos[2],c='r')
                
                ax.set_xlim(-10,10)
                ax.set_ylim(-10,10)
                ax.set_zlim(-10,10)
                
                fig.canvas.draw()
                
            else:
                return
        index = self.df.index[self.index]
      
        track = np.vstack([*self.df.loc[index,['XPosition','YPosition','ZPosition']].values])
        track = np.roll(np.append(track,np.array([[0.],[0.],[1500.]]),axis=1),1,axis=1)
        track = (track - (self.loc - 2*self.cs)[:,np.newaxis])/self.cs

        ax.plot(track[0],track[1],track[2])

        therm_pos = df.loc[index,['ThermalPosX','ThermalPosY','ThermalPosZ']].values.astype(float)
        therm_pos = (therm_pos - (self.loc - 2*self.cs))/self.cs

        ax.scatter(therm_pos[0],therm_pos[1],therm_pos[2],c='r')

        ax.set_xlim(-10,10)
        ax.set_ylim(-10,10)
        ax.set_zlim(-10,10)

        fig.canvas.mpl_connect('key_press_event',key_event)
                
                
    

In [151]:
wrap = Wrapper(df.loc[thermal.query('ThermalPosX<-100|ThermalPosX>100|ThermalPosY<-100|ThermalPosY>100|ThermalPosZ<-100|ThermalPosZ>100').index])

In [153]:
wrap.ViewArr3D()

In [157]:
wrap.df.index[wrap.index]

206

In [159]:
wrap.df.loc[206]

pdgid                                                          [2112]
energy                                          [0.10000000000002274]
posx                                                            [0.0]
posy                                                            [0.0]
posz                                                         [1500.0]
momx                                            [0.44713827970387204]
momy                                            [-0.8604151000636001]
momz                                            [0.24444470623436912]
DepositEnergy                                                4.881201
LiCaptureEnergy                                                   0.0
PvtCaptureEnergy                                                  0.0
CapPosX                                                     60.405299
CapPosY                                                  -1597.215728
CapPosZ                                                    1551.31518
ThermalPosX         

In [40]:
df.loc[49]['Energy']

array([1.00000000e+05, 9.46404552e+03, 3.04452065e+03, 2.32839158e+03,
       1.38009478e+02, 2.67331072e+01, 2.67331072e+01, 2.67331072e+01,
       1.27023585e+01, 9.62696981e+00, 9.52958453e+00, 9.52958453e+00,
       9.52958453e+00, 3.20816446e+00, 3.20816446e+00, 3.20816446e+00,
       3.20816446e+00, 3.20816446e+00, 1.68894098e+00, 1.68894098e+00,
       1.68894098e+00, 1.68894098e+00, 1.68894098e+00, 1.68894098e+00,
       1.04046304e+00, 1.04046304e+00, 1.04046304e+00, 1.04046304e+00,
       1.04046304e+00, 3.73116563e-01, 3.73116563e-01, 3.73116563e-01,
       1.64553806e-01, 1.60482099e-01, 1.60482099e-01, 1.60482099e-01,
       1.60482099e-01, 1.60482099e-01, 1.50697360e-01, 1.50697360e-01,
       1.50697360e-01, 1.70053115e-01, 1.19278411e-01, 1.19278411e-01,
       1.19278411e-01, 1.19278411e-01, 1.19278411e-01, 1.19278411e-01,
       1.19632788e-01, 1.11315768e-01, 1.11315768e-01, 1.11315768e-01,
       1.11315768e-01, 1.11315768e-01, 1.11315768e-01, 1.29130433e-01,
      

In [41]:
df.loc[49]['XPosition']

array([ 5.14412959e+01,  5.15534922e+01,  5.27981245e+01,  6.39659151e+01,
        6.74029302e+01,  6.71511302e+01,  6.98756669e+01,  1.20004500e+02,
        1.23494833e+02,  1.24871407e+02,  1.24292723e+02,  1.20004500e+02,
        7.83315891e+01,  7.41219726e+01,  7.24003178e+01,  7.23302800e+01,
        7.22689762e+01,  7.21826237e+01,  7.16935533e+01,  7.17644575e+01,
        7.17781516e+01,  7.17878734e+01,  7.17975952e+01,  7.18112894e+01,
        7.19652658e+01,  7.23244136e+01,  7.23804147e+01,  7.24201712e+01,
        7.24655919e+01,  7.30066129e+01,  7.01802670e+01,  3.12899187e+01,
        2.91618381e+01,  3.06271521e+01,  2.77757406e+01, -6.60829306e-01,
       -3.89798579e+00, -3.99625661e+00, -4.00251039e+00, -3.99851708e+00,
       -3.93576677e+00, -3.92516235e+00, -1.83781785e+00,  9.21145646e-03,
        9.36667252e-02,  1.86000000e-01,  2.71718304e-01,  4.48000000e-01,
        6.01638215e-01,  7.18090360e-01,  4.48000000e-01,  1.86000000e-01,
        5.82867088e-15, -

In [64]:
energies = np.sort([f for f in effs.keys() if type(f)==np.float64])
sources = [f for f in effs.keys() if type(f)==str]

cap_eff,cap_err = [],[]
scatt_eff,scatt_err = [],[]
scatt_1_eff,scatt_1_err = [],[]
scatt_2_eff,scatt_2_err = [],[]
cap_scatt_1_eff,cap_scatt_1_err = [],[]
cap_scatt_2_eff,cap_scatt_2_err = [],[]
caps,dfs = [],[]

for e in energies:
    cap_eff.append(100*effs[e]['cap_eff']), cap_err.append(100*errs[e]['cap_eff'])
    scatt_eff.append(100*effs[e]['scatt_eff']), scatt_err.append(100*errs[e]['scatt_eff'])
    scatt_1_eff.append(100*effs[e]['scatt_1_eff']), scatt_1_err.append(100*errs[e]['scatt_1_eff'])
    scatt_2_eff.append(100*effs[e]['scatt_2_eff']), scatt_2_err.append(100*errs[e]['scatt_2_eff'])
    cap_scatt_1_eff.append(100*effs[e]['cap_scatt_1_eff']), cap_scatt_1_err.append(100*errs[e]['cap_scatt_1_eff'])
    cap_scatt_2_eff.append(100*effs[e]['cap_scatt_2_eff']), cap_scatt_2_err.append(100*errs[e]['cap_scatt_2_eff'])
    caps.append(cap_lens[e])
    dfs.append(df_lens[e])
    
for s in sources:
    cap_eff.append(100*effs[s]['cap_eff']), cap_err.append(100*errs[s]['cap_eff'])
    scatt_eff.append(100*effs[s]['scatt_eff']), scatt_err.append(100*errs[s]['scatt_eff'])
    scatt_1_eff.append(100*effs[s]['scatt_1_eff']), scatt_1_err.append(100*errs[s]['scatt_1_eff'])
    scatt_2_eff.append(100*effs[s]['scatt_2_eff']), scatt_2_err.append(100*errs[s]['scatt_2_eff'])
    cap_scatt_1_eff.append(100*effs[s]['cap_scatt_1_eff']), cap_scatt_1_err.append(100*errs[s]['cap_scatt_1_eff'])
    cap_scatt_2_eff.append(100*effs[s]['cap_scatt_2_eff']), cap_scatt_2_err.append(100*errs[s]['cap_scatt_2_eff'])
    caps.append(cap_lens[s])
    dfs.append(df_lens[s])
    

In [91]:
energies

array([2.4999963e-08, 1.4400000e-01, 2.0000000e-01, 3.0000000e-01,
       4.0000000e-01, 5.0000000e-01, 6.0000000e-01, 7.0000000e-01,
       8.0000000e-01, 9.0000000e-01, 1.0000000e+00, 1.2000000e+00,
       2.0000000e+00, 5.0000000e+00])

In [11]:
cap_eff, cap_err

([66.36188907193849,
  40.8375314861461,
  39.10860012554927,
  40.57142857142857,
  39.80358358869594,
  37.2220530003046,
  35.218806441902466,
  37.542036074594925,
  38.479880774962744,
  38.59627558232438,
  36.43785908678561,
  35.96885294998503,
  35.9846244825547,
  37.66134774890429,
  35.52709946396665,
  36.84210526315789,
  34.754859297940236,
  36.6318358152922,
  33.69239976062238,
  34.943351222420986,
  34.761357101782636,
  36.428461449632394,
  35.3235294117647,
  35.28718703976436,
  36.855744313273824,
  35.856250893665695,
  33.82857142857143,
  32.95757935330762,
  34.068195179306294,
  35.22002414609337,
  34.37949899222574,
  34.69150174621653,
  33.37094499294781,
  34.817878310450894,
  33.285426846795055,
  33.201914953534214,
  33.90470656595003,
  34.65949753286376,
  33.60512521840419,
  33.83458646616541,
  33.53837141183363,
  33.60225776855878,
  30.648031373633017,
  29.449008614052406,
  20.29479277748658,
  12.436167017122258,
  8.976239366383105,
  

In [13]:
energies[4],caps[4],np.sqrt(caps[4]),dfs[4],np.sqrt(dfs[4]),cap_eff[4],cap_err[4]

(0.100000000000178,
 37409,
 193.4140636044856,
 93984,
 306.56810010175553,
 39.80358358869594,
 0.24332866677085)

In [14]:
energies[5],caps[5],np.sqrt(caps[5]),dfs[5],np.sqrt(dfs[5]),cap_eff[5],cap_err[5]

(0.125,
 1222,
 34.95711658589707,
 3283,
 57.29746940310715,
 37.2220530003046,
 1.2473165261773607)

In [57]:
def yield_vals(i):
    return energies[i],caps[i],np.sqrt(caps[i]),dfs[i],np.sqrt(dfs[i]),cap_eff[i],cap_err[i]

In [86]:
yield_vals(-3)

(10.0,
 25403,
 159.38318606427717,
 108768,
 329.7999393571806,
 23.35521476904972,
 0.16274967053994838)

In [106]:
energies

array([2.4999963e-08, 1.4400000e-01, 2.0000000e-01, 3.0000000e-01,
       4.0000000e-01, 5.0000000e-01, 6.0000000e-01, 7.0000000e-01,
       8.0000000e-01, 9.0000000e-01, 1.0000000e+00, 1.2000000e+00,
       2.0000000e+00, 5.0000000e+00])

In [108]:
it = iter(np.arange(len(energies)))

In [111]:
yield_vals(next(it))

(0.20000000000016382,
 25231,
 158.8426894760977,
 65045,
 255.03921267130667,
 38.7900684141748,
 0.2876951288876817)

##### NPL geometry

| Energy | N_entered | N_captured | Efficiency |
|--|--|--|--|
| 0.025 eV | 36420 $\pm$ 191 | 24169 $\pm$ 155 | 66.4 $\pm$ 0.6 % |
| 100 keV | 93984 $\pm$ 307 | 37409 $\pm$ 193 | 39.9 $\pm$ 0.2 % |
| 125 keV | 3283 $\pm$ 57 | 1222 $\pm$ 35 | 37.2 $\pm$ 1.2 % |
| 150 keV | 3271 $\pm$ 57 | 1228 $\pm$ 35 | 37.5 $\pm$ 1.3 % |
| 175 keV | 3355 $\pm$ 58 | 1291 $\pm$ 36 | 38.5 $\pm$ 1.3 % |
| 200 keV | 97626 $\pm$ 312 | 37680 $\pm$ 194 | 38.6 $\pm$ 0.2 % |


##### Empty geometry

| Energy | N_entered | N_captured | Efficiency |
|--|--|--|--|
|0.025 eV | 24473 $\pm$ 156 | 16289 $\pm$ 128 | 66.6 $\pm$ 0.7 % |
| 144 keV | 64427 $\pm$ 254 | 25448 $\pm$ 160 | 39.5 $\pm$ 0.3 % |
| 200 keV | 65045 $\pm$ 255 | 25231 $\pm$ 159 | 38.8 $\pm$ 0.3 % |




| 225 keV | 3307 $\pm$ 58 | 1205 $\pm$ 35 | 36.4 $\pm$ 1.2 % |
| 250 keV | 3339 $\pm$ 58 | 1201 $\pm$ 35 | 36.0 $\pm$ 1.2 % |
| 275 keV | 3382 $\pm$ 58 | 1217 $\pm$ 35 | 36.0 $\pm$ 1.2 % |
| 300 keV | 99707 $\pm$ 316 | 37551 $\pm$ 194 | 37.7 $\pm$ 0.2 % || 25 keV  | 3176 $\pm$ 56 | 1297 $\pm$ 36 | 40.8 $\pm$ 1.3 % |
| 50 keV  | 3186 $\pm$ 56 | 1246 $\pm$ 35 | 39.1 $\pm$ 1.3 % |
| 75 keV  | 3150 $\pm$ 56 | 1278 $\pm$ 36 | 40.6 $\pm$ 1.3 % |

In [4]:
dfs

[36420,
 3176,
 3186,
 3150,
 93984,
 3283,
 106304,
 3271,
 3355,
 97626,
 3307,
 3339,
 3382,
 99707,
 3358,
 3439,
 3447,
 95892,
 3342,
 3354,
 3478,
 103916,
 3400,
 3395,
 3473,
 104905,
 3500,
 3371,
 3402,
 105193,
 3473,
 3436,
 3545,
 106632,
 3479,
 3551,
 3442,
 106196,
 3434,
 3458,
 3414,
 100276,
 115766,
 110401,
 110654,
 3329,
 3409,
 3332,
 108768,
 50192,
 53756]

In [218]:
energies[4],energies[5]

(0.100000000000178, 0.125)

In [96]:
dose_eq_curves = pd.read_csv("/home/nr1315/Downloads/ICRP74_neutron_dose equivalent curves.csv")

In [17]:
import matplotlib.gridspec as gridspec

In [16]:
from scipy.optimize import curve_fit

In [18]:
def straight_line(x,m,c):
    return m*x + c

def exp_straight_line(x,m,c):
    return np.exp(straight_line(x,m,c))

def decay_exp(x,A,m,c,C):
    return A*np.exp(-1*m*x + c) + C

(m,c),sig = curve_fit(straight_line,np.log10(energies[:42]),cap_eff[:-3][:42],sigma = 2*np.array(cap_err)[:-3][:42])
(m2,c2),sig2 = curve_fit(exp_straight_line,energies[:42],cap_eff[:-3][:42],sigma=2*np.array(cap_err)[:-3][:42])
(A,m3,c3,C),sig3 = curve_fit(decay_exp,energies[:42],cap_eff[:-3][:42],sigma=2*np.array(cap_err)[:-3][:42])

In [26]:
inds = np.delete(np.arange(len(energies)),6)

In [56]:
(m1,c1),sig = curve_fit(straight_line,np.log10(energies[:42]),cap_eff[:-3][:42],sigma = 2*np.array(cap_err)[:-3][:42])
(m2,c2),sig = curve_fit(straight_line,np.log10(energies[inds][:41]),np.array(cap_eff)[:-3][inds][:41],sigma = 2*np.array(cap_err)[:-3][inds][:41])

#plt.plot(energies[inds][:41],np.array(cap_eff)[:-3][inds][:41],c='blue')
plt.plot(energies[:42],np.array(cap_eff)[:-3][:42],label = 'Simulated capture efficiency data',color='C0',linestyle='--')
plt.errorbar(energies[:42],np.array(cap_eff[:-3][:42]),yerr=np.array(cap_err[:-3][:42]),fmt='none',color='C0')
plt.fill_between(energies[:42],(np.array(cap_eff[:-3])-np.array(cap_err[:-3]))[:42],(np.array(cap_eff[:-3])+np.array(cap_err[:-3]))[:42],alpha=0.3,label='1 sigma region')

plt.plot(energies[inds][:41],straight_line(np.log10(energies[inds][:41]),m2,c2),label='Best fit line, without 144 keV',color='C1')
plt.plot(energies[:42],straight_line(np.log10(energies[:42]),m1,c1),label='Best fit line, all points',color='C2')
plt.legend(loc='upper right')
#plt.plot(energies[:42],exp_straight_line(energies[:42],m2,c2))
#plt.plot(energies[:42],decay_exp(energies[:42],A,m3,c3,C))
plt.xscale('log')
plt.xlabel('Energy / MeV',fontsize=24)
plt.ylabel('Capture Efficiency / \%',fontsize=24)

Text(0, 0.5, 'Capture Efficiency / \\%')

In [212]:
max_err = np.max(cap_err[:-3])

In [215]:
plt.plot(energies,cap_eff[:-3])
plt.fill_between(energies,np.array(cap_eff)[:-3] - 3*max_err,np.array(cap_eff[:-3])+3*max_err,alpha=0.3)
plt.xscale('log')

In [200]:
popt,pcov=curve_fit(straight_line,np.log10(e2[:41]),cap2[:41],sigma=err2[:41])

In [211]:
plt.plot(e2[:41],cap2[:41])
plt.plot(e2[:41],straight_line(np.log10(e2[:41]),*popt))
plt.xscale('log')
plt.plot(energies[:42],straight_line(np.log10(energies[:42]),m,c))


In [168]:
np.delete(energies,6)

array([2.4999963e-08, 2.5000000e-02, 5.0000000e-02, 7.5000000e-02,
       1.0000000e-01, 1.2500000e-01, 1.5000000e-01, 1.7500000e-01,
       2.0000000e-01, 2.2500000e-01, 2.5000000e-01, 2.7500000e-01,
       3.0000000e-01, 3.2500000e-01, 3.5000000e-01, 3.7500000e-01,
       4.0000000e-01, 4.2500000e-01, 4.5000000e-01, 4.7500000e-01,
       5.0000000e-01, 5.2500000e-01, 5.5000000e-01, 5.7500000e-01,
       6.0000000e-01, 6.2500000e-01, 6.5000000e-01, 6.7500000e-01,
       7.0000000e-01, 7.2500000e-01, 7.5000000e-01, 7.7500000e-01,
       8.0000000e-01, 8.2500000e-01, 8.5000000e-01, 8.7500000e-01,
       9.0000000e-01, 9.2500000e-01, 9.5000000e-01, 9.7500000e-01,
       1.0000000e+00, 1.2000000e+00, 2.0000000e+00, 5.0000000e+00,
       1.0000000e+01, 2.5000000e+01, 1.0000000e+02])

In [165]:
energies==(1.44e-1)

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False])

In [184]:
new_params = curve_fit(straight_line,np.log10(energies)[:42],cap_eff[:-3][:42],sigma=cap_err[:-3][:42])

In [185]:
plt.plot(np.log10(energies[:42]),cap_eff[:-3][:42])
plt.plot(np.log10(energies[:42]),straight_line(np.log10(energies[:42]),new_params[0][0],new_params[0][1]))

In [183]:
plt.plot(energies[:42],cap_eff[:-3][:42])
plt.xscale('log')

In [171]:
e2 = np.delete(energies,6)
cap2 = np.delete(cap_eff[:-3],6)
err2 = np.delete(cap_err[:-3],6)

(mp,cp),sigp = curve_fit(straight_line,e2[:42],cap2[:42],sigma = err2[:42])
(m2p,c2p),sig2p = curve_fit(exp_straight_line,e2[:42],cap2[:42],sigma=err2[:42])
(Ap,m3p,c3p,Cp),sig3p = curve_fit(decay_exp,e2[:42],cap2[:42],sigma=err2[:42])

plt.plot(e2[:42],cap2[:42])
plt.fill_between(e2[:42],(cap2-err2)[:42],(cap2+err2)[:42],alpha=0.3)
plt.plot(e2[:42],straight_line(e2[:42],mp,cp))
plt.plot(e2[:42],exp_straight_line(e2[:42],m2p,c2p))
plt.plot(e2[:42],decay_exp(e2[:42],Ap,m3p,c3p,Cp))
plt.xscale('log')

In [99]:
energies = efficiencies.index

In [103]:
type(energies[0])==np.float64

True

In [106]:
type(energies[6])==str

True

In [111]:
e = np.sort([i for i in energies if type(i)==np.float64])
s = [i for i in energies if type(i)==str]

In [67]:
dose_eq_curves = pd.read_csv("/home/nr1315/Downloads/ICRP74_neutron_dose equivalent curves.csv")

fig=plt.figure()
ax = fig.add_subplot(111)

ax.plot(e, efficiencies.loc[e,'cap_eff'], c = 'xkcd:tangerine', label = 'Simulated capture events', linestyle = '--')
ax.errorbar(e, efficiencies.loc[e,'cap_eff'], yerr = errors.loc[e,'cap_eff'], fmt = 'none', ecolor = 'xkcd:tangerine')

ax.fill_between(e,np.array(cap_eff[:-3])-3*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+3*np.array(cap_err[:-3]),color='xkcd:goldenrod',alpha=0.3)
ax.fill_between(e,np.array(cap_eff[:-3])-2*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+2*np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)
ax.fill_between(e,np.array(cap_eff[:-3])-np.array(cap_err[:-3]),np.array(cap_eff[:-3])+np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)

ax.plot(energies, scatt_eff[:-3], c = 'xkcd:deep blue', label = 'Simulated scatter events', linestyle = '--')
ax.errorbar(energies, scatt_eff[:-3], yerr = scatt_err[:-3], fmt = 'none', ecolor = 'xkcd:deep blue')

ax.fill_between(energies,np.array(scatt_eff[:-3])-3*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+3*np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff[:-3])-2*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+2*np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff[:-3])-np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)


#ax.plot(energies, scatt_1_eff, c = 'b', label = '1 detectable scatter')
#ax.errorbar(energies, scatt_1_eff, yerr = scatt_1_err, fmt = 'none', ecolor = 'b')

#ax.plot(energies, scatt_2_eff, c = 'xkcd:light blue', label = '2 detectable scatters')
#ax.errorbar(energies, scatt_2_eff, yerr = scatt_2_err, fmt = 'none', ecolor = 'xkcd:deep blue')

#ax.plot(energies, cap_scatt_1_eff, c = 'r', label = '1 scatter + capture')
#ax.errorbar(energies, cap_scatt_1_eff, yerr = cap_scatt_1_err, fmt = 'none', ecolor = 'r')

#ax.plot(energies, cap_scatt_2_eff, c = 'xkcd:dark red', label = '2 scatters + capture')
#ax.errorbar(energies, cap_scatt_2_eff, yerr = cap_scatt_2_err, fmt = 'none', ecolor = 'xkcd:dark red')

ax4 = ax.twinx()
ax4.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
ax4.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
ax4.tick_params(axis='y',labelcolor='g')

ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax.tick_params(axis='x',labelsize=18)

ax.tick_params(axis='y',labelsize=18)
ax.set_xlabel('Energy / MeV',fontsize=24)
ax.set_xlim(1e-8,1e2)
ax.set_xscale('log')

ax.scatter(real_es,100*np.array(real_effs),c='purple',label=r'Real capture events')
ax.errorbar(real_es,100*np.array(real_effs),yerr=100*np.array(real_errs),c='purple',fmt='none', capsize = 4)


ax.legend(loc='upper left')
fig.suptitle('Detection efficiency',fontsize=32)



NameError: name 'real_es' is not defined

In [9]:
#fig,(ax, ax2) = plt.subplots(1,2,sharey=True)
gs = gridspec.GridSpec(1,4)

fig = plt.figure()
ax = fig.add_subplot(gs[:,0])
ax2 = fig.add_subplot(gs[:,1:])


ax.plot(energies, cap_eff, c = 'xkcd:tangerine', label = 'All captured events', linestyle = '--')
ax.errorbar(energies, cap_eff, yerr = cap_err, fmt = 'none', ecolor = 'xkcd:tangerine')

ax.fill_between(energies,np.array(cap_eff)-3*np.array(cap_err),np.array(cap_eff)+3*np.array(cap_err),color='xkcd:goldenrod',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff)-2*np.array(cap_err),np.array(cap_eff)+2*np.array(cap_err),color='xkcd:light orange',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff)-np.array(cap_err),np.array(cap_eff)+np.array(cap_err),color='xkcd:light orange',alpha=0.3)

ax.plot(energies, scatt_eff, c = 'xkcd:deep blue', label = 'All scatter events', linestyle = '--')
ax.errorbar(energies, scatt_eff, yerr = scatt_err, fmt = 'none', ecolor = 'xkcd:deep blue')

ax.fill_between(energies,np.array(scatt_eff)-3*np.array(scatt_err),np.array(scatt_eff)+3*np.array(scatt_err),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff)-2*np.array(scatt_err),np.array(scatt_eff)+2*np.array(scatt_err),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff)-np.array(scatt_err),np.array(scatt_eff)+np.array(scatt_err),color='xkcd:light blue',alpha=0.3)


ax.plot(energies, scatt_1_eff, c = 'b', label = '1 detectable scatter')
ax.errorbar(energies, scatt_1_eff, yerr = scatt_1_err, fmt = 'none', ecolor = 'b')

ax.plot(energies, scatt_2_eff, c = 'xkcd:light blue', label = '2 detectable scatters')
ax.errorbar(energies, scatt_2_eff, yerr = scatt_2_err, fmt = 'none', ecolor = 'xkcd:deep blue')

ax.plot(energies, cap_scatt_1_eff, c = 'r', label = '1 scatter + capture')
ax.errorbar(energies, cap_scatt_1_eff, yerr = cap_scatt_1_err, fmt = 'none', ecolor = 'r')

ax.plot(energies, cap_scatt_2_eff, c = 'xkcd:dark red', label = '2 scatters + capture')
ax.errorbar(energies, cap_scatt_2_eff, yerr = cap_scatt_2_err, fmt = 'none', ecolor = 'xkcd:dark red')



ax2.plot(energies, cap_eff, c = 'xkcd:tangerine', label = 'All captured events', linestyle = '--')
ax2.errorbar(energies, cap_eff, yerr = cap_err, fmt = 'none', ecolor = 'xkcd:tangerine')
ax2.fill_between(energies,np.array(cap_eff)-3*np.array(cap_err),np.array(cap_eff)+3*np.array(cap_err),color='xkcd:goldenrod',alpha=0.3)
ax2.fill_between(energies,np.array(cap_eff)-2*np.array(cap_err),np.array(cap_eff)+2*np.array(cap_err),color='xkcd:light orange',alpha=0.3)
ax2.fill_between(energies,np.array(cap_eff)-np.array(cap_err),np.array(cap_eff)+np.array(cap_err),color='xkcd:light orange',alpha=0.3)
ax2.plot(energies, scatt_eff, c = 'xkcd:deep blue', label = 'All scatter events', linestyle = '--')
ax2.fill_between(energies,np.array(scatt_eff)-3*np.array(scatt_err),np.array(scatt_eff)+3*np.array(scatt_err),color='xkcd:baby blue',alpha=0.3)
ax2.fill_between(energies,np.array(scatt_eff)-2*np.array(scatt_err),np.array(scatt_eff)+2*np.array(scatt_err),color='xkcd:sky blue',alpha=0.3)
ax2.fill_between(energies,np.array(scatt_eff)-np.array(scatt_err),np.array(scatt_eff)+np.array(scatt_err),color='xkcd:cerulean blue',alpha=0.3)


ax2.errorbar(energies, scatt_eff, yerr = scatt_err, fmt = 'none', ecolor = 'xkcd:deep blue')
ax2.plot(energies, scatt_1_eff, c = 'b', label = '1 detectable scatter')
ax2.errorbar(energies, scatt_1_eff, yerr = scatt_1_err, fmt = 'none', ecolor = 'b')
ax2.plot(energies, scatt_2_eff, c = 'xkcd:light blue', label = '2 detectable scatters')
ax2.errorbar(energies, scatt_2_eff, yerr = scatt_2_err, fmt = 'none', ecolor = 'xkcd:light blue')
ax2.plot(energies, cap_scatt_1_eff, c = 'r', label = '1 scatter + capture')
ax2.errorbar(energies, cap_scatt_1_eff, yerr = cap_scatt_1_err, fmt = 'none', ecolor = 'r')
ax2.plot(energies, cap_scatt_2_eff, c = 'xkcd:dark red', label = '2 scatters + capture')
ax2.errorbar(energies, cap_scatt_2_eff, yerr = cap_scatt_2_err, fmt = 'none', ecolor = 'xkcd:dark red')

ax2.set_xscale('log')

ax4 = ax.twinx()
ax4.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
#ax4.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
#ax4.tick_params(axis='y',labelcolor='g')


ax3 = ax2.twinx()
ax3.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
ax3.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
ax3.tick_params(axis='y',labelcolor='g')


#ax2.set_xlabel('Energy / MeV',fontsize=24)
ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax2.legend(markerscale = 2, scatterpoints = 3, loc = 'upper left')
ax.set_xscale('log')
ax.set_xlim(1e-8,1e-7)
ax2.set_xlim(0.01,100)

ax.yaxis.tick_left()
#ax.tick_params(axis='y',which='both',right='False',labelright='off')
#ax4.tick_params(axis='y',which='both',right='False',labelright = 'off')

#ax2.tick_params(axis='y',which='both',right='False',left='False',labelright='off',labelleft='off')
ax3.yaxis.tick_right()
ax.spines['right'].set_visible(False)
ax3.spines['left'].set_visible(False)
ax2.spines['left'].set_visible(False)
ax4.spines['right'].set_visible(False)
ax4.set_yticks([])
ax4.set_yticks([],minor=True)
ax2.set_yticks([])
ax2.set_yticks([],minor=True)
#ax.minorticks_off()

d = .015 # how big to make the diagonal lines in axes coordinates
# arguments to pass plot, just so we don't keep repeating them
kwargs = dict(transform=ax.transAxes, color='k', clip_on=False)
ax.plot((1-d/2,1+d/2), (-d/2,+d/2), **kwargs)
ax.plot((1-d/2,1+d/2),(1-d/2,1+d/2), **kwargs)

kwargs.update(transform=ax2.transAxes)  # switch to the bottom axes
ax2.plot((-d/9,d/9), (1-d/2,1+d/2), **kwargs)
ax2.plot((-d/9,d/9), (-d/2,+d/2), **kwargs)

fig.suptitle('Simulated detection efficiency',fontsize=32)
fig.text(0.45, 0.04, 'Energy / MeV', va='center',fontsize=24)
ax.tick_params(axis='x',labelsize=18)
ax2.tick_params(axis='x',labelsize=18)
ax.tick_params(axis='y',labelsize=18)
ax2.tick_params(axis='y',labelsize=18)
ax3.tick_params(axis='y',labelsize=18)
ax.set_xticklabels([],minor=True)

NameError: name 'gridspec' is not defined

In [140]:
#fig,(ax, ax2) = plt.subplots(1,2,sharey=True)
gs = gridspec.GridSpec(1,4)

fig = plt.figure()
ax = fig.add_subplot(gs[:,0])
ax2 = fig.add_subplot(gs[:,1:])


ax.plot(energies, cap_eff[:-3], c = 'xkcd:tangerine', label = 'All captured events', linestyle = '--')
ax.errorbar(energies, cap_eff[:-3], yerr = cap_err[:-3], fmt = 'none', ecolor = 'xkcd:tangerine')

ax.fill_between(energies,np.array(cap_eff[:-3])-3*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+3*np.array(cap_err[:-3]),color='xkcd:goldenrod',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff[:-3])-2*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+2*np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff[:-3])-np.array(cap_err[:-3]),np.array(cap_eff[:-3])+np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)

ax.plot(energies, scatt_eff[:-3], c = 'xkcd:deep blue', label = 'All scatter events', linestyle = '--')
ax.errorbar(energies, scatt_eff[:-3], yerr = scatt_err[:-3], fmt = 'none', ecolor = 'xkcd:deep blue')

ax.fill_between(energies,np.array(scatt_eff[:-3])-3*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+3*np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff[:-3])-2*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+2*np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)
ax.fill_between(energies,np.array(scatt_eff[:-3])-np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+np.array(scatt_err[:-3]),color='xkcd:light blue',alpha=0.3)


ax.plot(energies, scatt_1_eff[:-3], c = 'b', label = '1 detectable scatter')
ax.errorbar(energies, scatt_1_eff[:-3], yerr = scatt_1_err[:-3], fmt = 'none', ecolor = 'b')

ax.plot(energies, scatt_2_eff[:-3], c = 'xkcd:light blue', label = '2 detectable scatters')
ax.errorbar(energies, scatt_2_eff[:-3], yerr = scatt_2_err[:-3], fmt = 'none', ecolor = 'xkcd:deep blue')

ax.plot(energies, cap_scatt_1_eff[:-3], c = 'r', label = '1 scatter + capture')
ax.errorbar(energies, cap_scatt_1_eff[:-3], yerr = cap_scatt_1_err[:-3], fmt = 'none', ecolor = 'r')

ax.plot(energies, cap_scatt_2_eff[:-3], c = 'xkcd:dark red', label = '2 scatters + capture')
ax.errorbar(energies, cap_scatt_2_eff[:-3], yerr = cap_scatt_2_err[:-3], fmt = 'none', ecolor = 'xkcd:dark red')



ax2.plot(energies, cap_eff[:-3], c = 'xkcd:tangerine', label = 'All captured events', linestyle = '--')
ax2.errorbar(energies, cap_eff[:-3], yerr = cap_err[:-3], fmt = 'none', ecolor = 'xkcd:tangerine')
ax2.fill_between(energies,np.array(cap_eff[:-3])-3*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+3*np.array(cap_err[:-3]),color='xkcd:goldenrod',alpha=0.3)
ax2.fill_between(energies,np.array(cap_eff[:-3])-2*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+2*np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)
ax2.fill_between(energies,np.array(cap_eff[:-3])-np.array(cap_err[:-3]),np.array(cap_eff[:-3])+np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)
ax2.plot(energies, scatt_eff[:-3], c = 'xkcd:deep blue', label = 'All scatter events', linestyle = '--')
ax2.fill_between(energies,np.array(scatt_eff[:-3])-3*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+3*np.array(scatt_err[:-3]),color='xkcd:baby blue',alpha=0.3)
ax2.fill_between(energies,np.array(scatt_eff[:-3])-2*np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+2*np.array(scatt_err[:-3]),color='xkcd:sky blue',alpha=0.3)
ax2.fill_between(energies,np.array(scatt_eff[:-3])-np.array(scatt_err[:-3]),np.array(scatt_eff[:-3])+np.array(scatt_err[:-3]),color='xkcd:cerulean blue',alpha=0.3)


ax2.errorbar(energies, scatt_eff[:-3], yerr = scatt_err[:-3], fmt = 'none', ecolor = 'xkcd:deep blue')
ax2.plot(energies, scatt_1_eff[:-3], c = 'b', label = '1 detectable scatter')
ax2.errorbar(energies, scatt_1_eff[:-3], yerr = scatt_1_err[:-3], fmt = 'none', ecolor = 'b')
ax2.plot(energies, scatt_2_eff[:-3], c = 'xkcd:light blue', label = '2 detectable scatters')
ax2.errorbar(energies, scatt_2_eff[:-3], yerr = scatt_2_err[:-3], fmt = 'none', ecolor = 'xkcd:light blue')
ax2.plot(energies, cap_scatt_1_eff[:-3], c = 'r', label = '1 scatter + capture')
ax2.errorbar(energies, cap_scatt_1_eff[:-3], yerr = cap_scatt_1_err[:-3], fmt = 'none', ecolor = 'r')
ax2.plot(energies, cap_scatt_2_eff[:-3], c = 'xkcd:dark red', label = '2 scatters + capture')
ax2.errorbar(energies, cap_scatt_2_eff[:-3], yerr = cap_scatt_2_err[:-3], fmt = 'none', ecolor = 'xkcd:dark red')

ax2.set_xscale('log')

ax4 = ax.twinx()
ax4.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
#ax4.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
#ax4.tick_params(axis='y',labelcolor='g')


ax3 = ax2.twinx()
ax3.plot(dose_eq_curves['En / MeV'].values,dose_eq_curves['H*(10) / pSv.cm2'],c='g')
ax3.set_ylabel(r'$H^*(10)$ / pSv.cm$^2$',fontsize=24,c='g')
ax3.tick_params(axis='y',labelcolor='g')


#ax2.set_xlabel('Energy / MeV',fontsize=24)
ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax2.legend(markerscale = 2, scatterpoints = 3, loc = 'upper left')
ax.set_xscale('log')
ax.set_xlim(5e-9,1e-7)
ax2.set_xlim(0.01,100)

ax.yaxis.tick_left()
#ax.tick_params(axis='y',which='both',right='False',labelright='off')
#ax4.tick_params(axis='y',which='both',right='False',labelright = 'off')

#ax2.tick_params(axis='y',which='both',right='False',left='False',labelright='off',labelleft='off')
ax3.yaxis.tick_right()
ax.spines['right'].set_visible(False)
ax3.spines['left'].set_visible(False)
ax2.spines['left'].set_visible(False)
ax4.spines['right'].set_visible(False)
ax4.set_yticks([])
ax4.set_yticks([],minor=True)
ax2.set_yticks([])
ax2.set_yticks([],minor=True)

d = .015 # how big to make the diagonal lines in axes coordinates
# arguments to pass plot, just so we don't keep repeating them
kwargs = dict(transform=ax.transAxes, color='k', clip_on=False)
ax.plot((1-d/2,1+d/2), (-d/2,+d/2), **kwargs)
ax.plot((1-d/2,1+d/2),(1-d/2,1+d/2), **kwargs)

kwargs.update(transform=ax2.transAxes)  # switch to the bottom axes
ax2.plot((-d/9,d/9), (1-d/2,1+d/2), **kwargs)
ax2.plot((-d/9,d/9), (-d/2,+d/2), **kwargs)

fig.suptitle('Simulated detection efficiency',fontsize=32)
fig.text(0.45, 0.04, 'Energy / MeV', va='center',fontsize=24)
ax.tick_params(axis='x',labelsize=18)
ax2.tick_params(axis='x',labelsize=18)
ax.tick_params(axis='y',labelsize=18)
ax2.tick_params(axis='y',labelsize=18)
ax3.tick_params(axis='y',labelsize=18)


ValueError: x and y must have same first dimension, but have shapes (19,) and (16,)

## Simulated detection efficiency

![Sim detection efficiency](Plots/Sim_detection_eff.png)

It is important to note that the point for the capture efficiency at thermal energies cannot be easily shown on this graph due to the scale (otherwise it is difficult to see the variation at high energy), but it is ~70% at 0.025 eV. 

There are a few key takeaways from this plot:
* Capture efficiency decreases as a function of energy reasonably slowly until about 1 MeV
* Scatter efficiency rises quickly from the scatter detection threshold of 480 keV up to about 60% in the 2-5 MeV range, until it decreases rapidly with increasing energy as neutrons have sufficient energy to pass through the detector without scattering
* For measuring neutrons that both scatter and capture the best efficiency is in the 1-5 MeV region, as the 1 scatter + capture efficiency decreases as the 2 scatter + capture becomes measureable.

['AmBe, avg energy 4.210736158652032',
 'AmLi, avg energy 0.5531349146160244',
 'Cf252, avg energy 2.1758734457551316']

In [11]:
colours = ['green','blue','red']
source_e = [float(s.split(' ')[-1]) for s in sources]
stds = list(stds.values())


In [4]:
fluence_rates = {'144 keV':1.399,
                 '144 keV shadowcone':1.378,
                 '144 keV -30deg':1.395,
                 '144 keV -45deg':1.392,
                 '144 keV -60deg':1.361,
                 '144 keV -90deg':1.359,
                 '144 keV -135deg':1.337,
                 '144 keV -135deg shadowcone':1.503,
                 '144 keV -180deg':1.432,
                 '144 keV -210deg':1.414,
                 '144 keV -225d eg':1.412,
                 '144 keV -240deg':1.415,
                 '144 keV -240deg shadowcone':1.392,
                 '144 keV -270deg':1.379,
                 '144 keV -315deg':1.369,
                 '1.2 MeV':1.664,
                 '1.2 MeV shadowcone':1.694,
                 '1.2 MeV -30deg':1.683,
                 '1.2 MeV -30deg shadowcone':1.595,
                 '1.2 MeV -45deg':1.552,
                 '1.2 MeV -45deg shadowcone':1.478,
                 '1.2 MeV -60deg':1.540,
                 '1.2 MeV -60deg shadowcone':1.544}

fl_uncertainty = {'144 keV':0.071, '1.2 MeV':0.06}


In [5]:
real_effs = []
real_errs = []
real_es = [0.144,1.2]

rv = RVS.RealVsSimAnalysis()
rv.add_real_data('/home/nr1315/Documents/Project/NPL_2021/vdg_1um_Li7target19A_433cm_0deg.root','144 keV')
rv.add_real_data('/home/nr1315/Documents/Project/NPL_2021/vdg_1um_Li7target19A_433cm_0deg_shadowcone.root','144 keV shadowcone')
sub_r,sub_r_err,real_r,real_r_err = rv.get_sdc_subtracted('144 keV','144 keV shadowcone',fluence_rates,fl_uncertainty)
sub_r = sub_r.sum()
sub_r_err = np.sqrt((sub_r_err**2).sum())

real_effs.append(sub_r/real_r)
real_errs.append(real_effs[0]*np.sqrt((sub_r_err**2/sub_r**2 + real_r_err**2/real_r**2)))


rv = RVS.RealVsSimAnalysis()
rv.add_real_data('/home/nr1315/Documents/Project/NPL_2021/vdg_1um_1_2_MeV_target3_2.root','1.2 MeV')
rv.add_real_data('/home/nr1315/Documents/Project/NPL_2021/vdg_1um_1_2_MeV_target3_shadowcone.root','1.2 MeV shadowcone')
sub_r,sub_r_err,real_r,real_r_err = rv.get_sdc_subtracted('1.2 MeV','1.2 MeV shadowcone',fluence_rates,fl_uncertainty)
sub_r = sub_r.sum()
sub_r_err = np.sqrt((sub_r_err**2).sum())

real_effs.append(sub_r/real_r)
real_errs.append(real_effs[1]*np.sqrt((sub_r_err**2/sub_r**2 + real_r_err**2/real_r**2)))

Finished loading /home/nr1315/Documents/Project/NPL_2021/vdg_1um_Li7target19A_433cm_0deg.root
------------------------------
Finished loading /home/nr1315/Documents/Project/NPL_2021/vdg_1um_Li7target19A_433cm_0deg_shadowcone.root
------------------------------
Finished loading /home/nr1315/Documents/Project/NPL_2021/vdg_1um_1_2_MeV_target3_2.root
------------------------------
Finished loading /home/nr1315/Documents/Project/NPL_2021/vdg_1um_1_2_MeV_target3_shadowcone.root
------------------------------


In [148]:
fig = plt.figure()
ax = fig.add_subplot(111)

ax.plot(energies, cap_eff[:-3], c = 'xkcd:tangerine', label = 'All captured events', linestyle = '--')
ax.errorbar(energies, cap_eff[:-3], yerr = cap_err[:-3], fmt = 'none', ecolor = 'xkcd:tangerine')

ax.fill_between(energies,np.array(cap_eff[:-3])-3*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+3*np.array(cap_err[:-3]),color='xkcd:goldenrod',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff[:-3])-2*np.array(cap_err[:-3]),np.array(cap_eff[:-3])+2*np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)
ax.fill_between(energies,np.array(cap_eff[:-3])-np.array(cap_err[:-3]),np.array(cap_eff[:-3])+np.array(cap_err[:-3]),color='xkcd:light orange',alpha=0.3)


for s in range(3):
    ax.scatter(source_e[s],cap_eff[(s-3)],c = colours[s], label = sources[s].split(',')[0])
    ax.errorbar(source_e[s],cap_eff[(s-3)],xerr = np.array(stds[s]),yerr = cap_err[(s-3)],c=colours[s],fmt='none', capsize = 4)

    
    
ax.scatter(real_es,100*np.array(real_effs),c='purple',label='Real efficiency')
ax.errorbar(real_es,100*np.array(real_effs),yerr=100*np.array(real_errs),c='purple',fmt='none', capsize = 4)
    
ax.set_xlabel('Energy / MeV',fontsize=24)
ax.set_ylabel('Detection efficiency / $\%$',fontsize=24)
ax.legend(markerscale = 1, scatterpoints = 1, loc = 'upper right')
ax.set_xscale('log')
ax.set_xlim(0.01,100)

plt.title('Capture efficiency',fontsize=32)

ValueError: x and y must have same first dimension, but have shapes (19,) and (16,)

![Real vs sim efficiency](Plots/real_vs_sim_efficiency.png)

The point in energy for the AmBe, AmLi and $^{252}$Cf data points is the mean energy of the source, and the energy error bars show the interquartile range of the energy distribution for those sources. There are a few points of note in this plot:

* For the AmBe, AmLi, and $^{252}$Cf sources, their efficiency is generally close to the efficiency at their mean energy, with an increasing departure from the monoenergetic efficiency for the higher mean energy sources (namely $^{252}$Cf and AmBe sources). This is likely due to the lower energy component of these sources having a higher efficiency, whilst for AmLi the monoenergetic efficiency undergoes much less change over the energy region of interest.

* For the real data point at 144 keV, the simulated efficiency $\epsilon_{sim}$ is greater than the real efficiency $\epsilon_{real}$ as expected. If we compute the ratio of these two efficiencies, we find that $\epsilon_{real} \simeq 0.85\epsilon_{sim}$, as seen in the extraction of the scattered component for this dataset (**TO BE ADDED**)

* For the real data point at 1.2 MeV, the simulated efficiency is actually less than the real efficiency, which is obviously counter to expectations. This is likely due to neutrons that pass through or scatter out of the detector that but then scatter back in off of the back wall in the NPL facility, thus increasing the real efficiency. In contrast, in the simulation the detector is in a much more open space and as such any neutrons that pass through the detector will not scatter back in in the same way, reducing the simulated efficiency. In order to verify this, the simulation at 1.2 Mev should be repeated with the detector in the simulated NPL geometry, in the same position as in the real dataset. 